### ISSUE TITLE
Problem deleting HIERARCH keywords

### ISSUE THREAD

**Issue by alexrudy** [astropy/astropy#3687]
Astropy seems unable to delete a HIERARCH keyword:

```
from astropy.io import fits
header = fits.Header()
header['a(bc)d'] = 10
print(header['a(bc)d'])
del header['a(bc)d']
```

The last line raises a value error when trying to remove the index from a list stored in `_keyword_indices`.

I don't think I understand enough about how HIERARCH keywords are handled in astropy (I looked around in the Header class) to fix this... I think it is making an incorrect assumption about the contents of `Header._keyword_indices`, but I can't tell what to do.

Found this in astropy 1.0.1

**Comment by embray** [comment]
Confirmed; there's definitely something fishy going on there.

**Comment by embray** [comment]
Not that it should matter, but according to http://fits.gsfc.nasa.gov/registry/hierarch/hierarch.pdf

> Under the ESO implementation of this convention, each token string that precedes the
equals sign must only contain characters that are legal in formal FITS keywords, i.e., the
uppercase letters A through Z, the digits 0 through 9, and the hyphen and underscore
characters. The tokens may, however, be longer than the 8 character limit of formal FITS
keywords.

However, historically pyfits, and by extension astropy, have allowed any printable ASCII characters in HIERARCH keywords.  I'm not sure why this is.  The linked document describing the convention came out in 2009, but the convention has been around since will before then, so maybe it was previously more loosely-defined.  

(No, there's no reason to have versions for FITS formats, none whatsoever--all FITS is the same FITS =_=)

**Comment by pllim** [comment]
FITS likes to throw fits?

**Comment by alexrudy** [comment]
@embray Valid! 

Honestly, I'm fine if the solution is that I shouldn't have ever been able to set that keyword to begin with. In fact, thats what happened, I stuck an illegal character into the keyword (forgetting which ones are legal) saved, reloaded, thought "Hmm, thats a bad idea, better delete it..." and hit this snag.

**Comment by embray** [comment]
I found a quick fix for this in the code, but unfortunately it broke some other things.  I need to rethink a bit more about how HIERARCH cards are handled (there are so many ambiguities).  In the meantime it seems the best workaround is to take two slices of the header--before and after the bad keyword--and then append them to each other :/

**Comment by embray** [comment]
@alexrudy The bug here though didn't ultimately have to do with the fact that it contained bad characters since the code is, in principle, trying to support HIERARCH keywords with arbitrary characters.  The real problem comes down to ambiguity about case-sensitivity.

**Comment by saimn** [comment]
I cannot reproduce, I bet it was fixed by #5321

**Comment by embray** [comment]
Yes, this would have been fixed by #5321.
